include/npcomp/RefBackend/Runtime/UserAPI.h

```diff
@@ -150,6 +154,131 @@ class Tensor {
+  _(None)                                                                      \
+  _(Bool)                                                                      \
+  _(Int)                                                                       \
+  _(Double)                                                                    \
+  _(Tensor)
+
+struct RuntimeValue final {
+
+  RuntimeValue() : payload{0}, tag(Tag::None) {}
+
+  // Bool
+  RuntimeValue(bool b) : tag(Tag::Bool) { payload.asBool = b; }
+  bool isBool() const { return Tag::Bool == tag; }
+  bool toBool() const {
+    assert(isBool());
+    return payload.asBool;
+  }
+
+  // Int
+  RuntimeValue(std::int64_t i) : tag(Tag::Int) { payload.asInt = i; }
+  RuntimeValue(std::int32_t i) : RuntimeValue(static_cast<int64_t>(i)) {}
+  bool isInt() const { return Tag::Int == tag; }
+  bool toInt() const {
+    assert(isInt());
+    return payload.asInt;
+  }
+
+  // Double
+  RuntimeValue(double d) : tag(Tag::Double) { payload.asDouble = d; }
+  bool isDouble() const { return Tag::Double == tag; }
+  bool toDouble() const {
+    assert(isDouble());
+    return payload.asDouble;
+  }
+
+  // Tensor
+  RuntimeValue(Ref<Tensor> tensor) : tag(Tag::Tensor) {
+    payload.asRefPtr = static_cast<void *>(tensor.takePtr());
+  }
+  bool isTensor() const { return Tag::Tensor == tag; }
+  Ref<Tensor> toTensor() const {
+    assert(isTensor());
+    return Ref<Tensor>(reinterpret_cast<Tensor *>(payload.asRefPtr));
+  }
+
+  // RuntimeValue (downcast)
+  const RuntimeValue &toRuntimeValue() const { return *this; }
+  RuntimeValue &toRuntimeValue() { return *this; }
+
+  // Stringify tag for debugging.
+  std::string tagKind() const {
+    switch (tag) {
+#define DEFINE_CASE(x)                                                         \
+  case Tag::x:                                                                 \
+    return #x;
+      NPCOMP_FORALL_TAGS(DEFINE_CASE)
+#undef DEFINE_CASE
+    }
+    // TODO(brycearden): Print tag here
+    return "InvalidTag!";
+  }
+
+  RuntimeValue(const RuntimeValue &rhs) : RuntimeValue(rhs.payload, rhs.tag) {
+    if (isTensor()) {
+      reinterpret_cast<Tensor *>(payload.asRefPtr)->refCount += 1;
+    }
+  }
+  RuntimeValue(RuntimeValue &&rhs) noexcept : RuntimeValue() { swap(rhs); }
+
+  RuntimeValue &operator=(RuntimeValue &&rhs) & noexcept {
+    RuntimeValue(std::move(rhs)).swap(*this); // this also sets rhs to None
+    return *this;
+  }
+  RuntimeValue &operator=(RuntimeValue const &rhs) & {
+    RuntimeValue(rhs).swap(*this);
+    return *this;
+  }
+
+  ~RuntimeValue() {
+    if (isTensor()) {
+      Tensor *tensor = reinterpret_cast<Tensor *>(payload.asRefPtr);
+      if (tensor->refCount) {
+        tensor->refCount -= 1;
+      } else {
+        assert(false && "Expected a non-zero refCount for RuntimeValue");
+      }
+
+      // Let RAII go out of scope and call the destructor on Ref
+      Ref<Tensor> unused(tensor);
+    }
+  }
```

Comment:
There are some dragons here to make sure the `refCount` on the underlying `Tensor` object gets decremented properly.

Reply:
Yeah, this stuff is tricky. I think with RefTarget, it should be close-enough to IValue that we don't have to think too hard about it (just copy the idioms used there).